txp/Pos/Txp/Toil/Types.hs

```diff
@@ -124,18 +127,41 @@ instance Default MemPool where
         }
 
 ----------------------------------------------------------------------------
--- ToilModifier
+-- UtxoModifier, UndoMap and AddrCoinsMap
 ----------------------------------------------------------------------------
 
 type UtxoModifier = MM.MapModifier TxIn TxOutAux
 type UndoMap = HashMap TxId TxUndo
-
-fromUtxo :: Utxo -> UtxoModifier
-fromUtxo = foldr (uncurry MM.insert) mempty . M.toList
+type AddrCoinMap = HashMap Address Coin
+
+utxoToModifier :: Utxo -> UtxoModifier
+utxoToModifier = foldr (uncurry MM.insert) mempty . M.toList
+
+-- | Takes utxo modifier and address-coin map with correspodning utxo
+-- and applies utxo modifier to map.
+-- Works for O(size of modifier * log (size of map)).
+applyUtxoModToAddrCoinMap
+    :: UtxoModifier
+    -> (AddrCoinMap, Utxo)
+    -> AddrCoinMap
+applyUtxoModToAddrCoinMap modifier (addrCoins, utxo) =
+    let outToPair = view _TxOut . toaOut in
+    let resolvedAddrs =
+          mapMaybe (fmap outToPair . flip M.lookup utxo)
+                   (MM.deletions modifier) in
+    let subAddress r c = if r < c then Just (c `unsafeSubCoin` r) else Nothing in
+    let updateHM (ad, coins) = HM.update (subAddress coins) ad in
+    let addrCoinsRest = foldr updateHM addrCoins resolvedAddrs in
+    let modAddrCoins = map (outToPair . snd) (MM.insertions modifier) in
+    foldr (uncurry $ HM.insertWith unsafeAddCoin) addrCoinsRest modAddrCoins
```

Comment:
I find a bit hard to decompress this function in my head, mostly due to the use of let bindings coupled with folds. Maybe splitting this function into a bunch of `where` clauses, each one describing what is meant to be doing might help? In particular, as we are essentially transforming an `AddrCoinMap` (as we get as input an `AddrCoinMap` and we produce as output another `AddrCoinMap`) cannot we exploit the rich API `containers` give us to get the job done?

I've also noticed how lazy this function is (in particular, any operation on maps which is not `seq`-ed appropriately will end up building thunks), which I'm wondering if could bite us in the back down the line (of course not always the solution is to go _strict_, so without any tangible measurement I really cannot tell 😁 ) .